pkg/proxy/conntrack/cleanup.go

```diff
@@ -20,91 +20,105 @@ limitations under the License.
+	// serviceNodePortEndpointIPs maps service NodePort to the set of serving endpoint IPs.
+	serviceNodePortEndpointIPs := make(map[int]sets.Set[string])
+
+	for svcName, svc := range svcPortMap {
+		// we are only interested in UDP services
+		if svc.Protocol() != v1.ProtocolUDP {
+			continue
+		}
+
+		endpointIPs := sets.New[string]()
+		for _, endpoint := range endpointsMap[svcName] {
+			// We need to remove all the conntrack entries for a Service (IP or NodePort)
+			// that are not pointing to a serving endpoint.
+			// We map all the serving endpoint IPs to the service and clear all the conntrack
+			// entries which are destined for the service and are not DNATed to these endpoints.
+			// Changes to the service should not affect existing flows, so we do not take
+			// traffic policies, topology, or terminating status of the service into account.
+			// This ensures that the behavior of UDP services remains consistent with TCP
+			// services.
+			if endpoint.IsServing() {
+				endpointIPs.Insert(endpoint.IP())
+			}
+		}
+
+		serviceIPEndpointIPs[svc.ClusterIP().String()] = endpointIPs
+		for _, loadBalancerIP := range svc.LoadBalancerVIPs() {
+			serviceIPEndpointIPs[loadBalancerIP.String()] = endpointIPs
+		}
+		for _, externalIP := range svc.ExternalIPs() {
+			serviceIPEndpointIPs[externalIP.String()] = endpointIPs
+		}
+		if svc.NodePort() != 0 {
+			serviceNodePortEndpointIPs[svc.NodePort()] = endpointIPs
+		}
 	}
-}
 
-// deleteStaleEndpointConntrackEntries takes care of flushing stale conntrack entries related
-// to UDP endpoints. After a UDP endpoint is removed we must flush any conntrack entries
-// for it so that if the same client keeps sending, the packets will get routed to a new endpoint.
-func deleteStaleEndpointConntrackEntries(ct Interface, ipFamily v1.IPFamily, svcPortMap proxy.ServicePortMap, endpointsUpdateResult proxy.UpdateEndpointsMapResult) {
 	var filters []netlink.CustomConntrackFilter
-	for _, epSvcPair := range endpointsUpdateResult.DeletedUDPEndpoints {
-		if svcInfo, ok := svcPortMap[epSvcPair.ServicePortName]; ok {
-			endpointIP := proxyutil.IPPart(epSvcPair.Endpoint)
-			nodePort := svcInfo.NodePort()
-			if nodePort != 0 {
-				filters = append(filters, filterForPortNAT(endpointIP, nodePort, v1.ProtocolUDP))
+	for _, entry := range entries {
+		// we only deal with UDP protocol entries
+		if entry.Forward.Protocol != unix.IPPROTO_UDP {
+			continue
+		}
 
+		origDst := entry.Forward.DstIP.String()
+		origPortDst := int(entry.Forward.DstPort)
+		replySrc := entry.Reverse.SrcIP.String()
+
+		// if the original destination (--orig-dst) of the entry is service IP (ClusterIP,
+		// LoadBalancerIPs or ExternalIPs) and the reply source (--reply-src) is not IP of
+		// any serving endpoint, we clear the entry.
+		if _, ok := serviceIPEndpointIPs[origDst]; ok {
+			if !serviceIPEndpointIPs[origDst].Has(replySrc) {
+				filters = append(filters, filterForNAT(origDst, replySrc, v1.ProtocolUDP))
```

Comment:
This is my fault, I do not know why I didn't catch this, we MUST filter but IP:Port destination, we only care about the existing services that are working, not about any Service that has may be existed, and we have this information on the informer cache ... sorry about it